cdap-app-fabric/src/main/java/co/cask/cdap/metadata/MetadataHttpHandler.java

```diff
@@ -215,6 +215,34 @@ public void removeTag(HttpRequest request, HttpResponder responder,
                          String.format("Metadata tag %s for %s deleted successfully.", tag, metadataEntity));
   }
 
+  @GET
+  @Path("/metadata/search")
+  public void searchMetadata(HttpRequest request, HttpResponder responder,
+                             @Nullable @QueryParam("query") String searchQuery,
+                             @Nullable @QueryParam("target") List<String> targets,
+                             @QueryParam("sort") @DefaultValue("") String sort,
+                             @QueryParam("offset") @DefaultValue("0") int offset,
+                             // 2147483647 is Integer.MAX_VALUE
+                             @QueryParam("limit") @DefaultValue("2147483647") int limit,
+                             @QueryParam("numCursors") @DefaultValue("0") int numCursors,
+                             @QueryParam("cursor") @DefaultValue("") String cursor,
+                             @QueryParam("showHidden") @DefaultValue("false") boolean showHidden,
+                             @QueryParam("showCustom") boolean showCustom,
+                             @Nullable @QueryParam("entityScope") String entityScope) throws Exception {
+    SearchRequest searchRequest = getValidatedSearchRequest(null, searchQuery, targets, sort, offset,
```

Comment:
One small change: Can we define a ALL constant or something in NamespaceId which represent all namespaces so that we don't have to pass around null ?

Reply:
"all" is a valid namespace though.

Reply:
I can make the getters return Optionals though